### PR TITLE
[RFC] Related content

### LINKED ISSUE #192: More elastic relations between content types
(Multiple relations to same content type but with different names or at least multiple sortable and limitable)

### PR DESCRIPTION
**Pull request by JarJak**
Closes #192 

Features:
---
- relations are decoupled from Content
- named relations allowed
- multi contenttype relations allowed
- sortable relations allowed
- relations are bidirectional by default, but owning side takes precedence

Configuration:
---
Config (from contentypes.yml):
```yaml
entries:
    relations:
        related_pages_and_blogposts: # name of relation, defaults to related content type if to_contenttypes is not set
            to_contenttypes: [pages, blogposts] # optional, see above
            sortable: true # default: false
            multiple: true # default: false
            min: 1
            max: 5
```
Or just simple config with defaults:
```yaml
entries:
    relations:
        pages: ~
        blogposts: ~
```

Twig examples:
---
```twig
{{ record|related(ct) }} -> Content[]
{{ record|related(name, bidirectional=false, limit=10) }} -> Content[]
{{ record|related_first(name) }} -> Content|null
{{ record|related_all }} or {{ related_content_all(record) }} -> Array<relation-name><Content>
```

API examples:
---
TBD: I thought about something like `/related/{from-id}/{to-contenttype}`

TODO:
---
- [x] write examples in skeleton
- [x] fixtures
- [x] readme
- [x] api endpoints
- [x] api tests with examples
- [x] vue components for admin
- [x] admin e2e tests
- [x] relations in preview mode
- [ ] rebase to master
- [x] split #163

### CODE REVIEW

**Comment by JarJak** [comment]
Will be rebased to master after https://github.com/bolt/four/pull/317 gets merged.

File: package.json [hunk 1]
```diff
@@ -65,13 +65,13 @@
...
     "babel-eslint": "^10.0.1",
-    "babel-jest": "^23.6.0",
+    "babel-jest": "^24.1.0",
     "eslint": "5.12.0",
     "eslint-config-prettier": "^3.6.0",
     "eslint-plugin-prettier": "^3.0.1",
     "eslint-plugin-standard": "^4.0.0",
     "eslint-plugin-vue": "^5.1.0",
-    "jest": "^23.6.0",
+    "jest": "^24.1.0",
```
Comment (bobdenotter) [inline package.json]: After we merge in this PR, #336 will give a conflict. I'll fix that one after this one is merged.

File: public/theme/skeleton/partials/_recordfooter.twig [hunk 2]
```diff
@@ -34,13 +34,19 @@
         </p>
     {% endif %}
 
-    {# @todo See Github Issue https://github.com/bolt/four/issues/252 #}
-    {% set relatedrecords = related_content(record) %}
-    {% if relatedrecords is not empty %}
+    {% set related_cts = record|related_all %}
```
Comment (bobdenotter) [inline public/theme/skeleton/partials/_recordfooter.twig]: Not sure if I like `related_cts`. Why not simply use `related_content` ?

File: src/Entity/Content.php [hunk 3]
```diff
@@ -46,7 +50,7 @@ class Content implements \JsonSerializable
     /**
      * @var string
      *
-     * @ORM\Column(type="string", length=191, name="contenttype")
+     * @ORM\Column(type="string")
```
Comment (bobdenotter) [inline src/Entity/Content.php]: Removing the `length=191` will break on MySQL 5.6. See https://github.com/symfony/symfony-docs/issues/6398

File: src/Entity/Content.php [hunk 4]
```diff
@@ -63,15 +67,15 @@ class Content implements \JsonSerializable
     /**
      * @var string|null
      *
-     * @ORM\Column(type="string", length=191)
+     * @ORM\Column(type="string")
```
Comment (bobdenotter) [inline src/Entity/Content.php]: Same here

File: src/Entity/Content.php (hunk 3 above)
Comment (JarJak) [reply src/Entity/Content.php]: Thanks, didn't know that! But this happens only on columns with `unique=true`, right? https://github.com/symfony/symfony-docs/pull/10055/files

Comment (bobdenotter) [reply src/Entity/Content.php]: AFAIK it happens on all columns that are either `unique`, used as an index or as a Foreign Key.

File: public/theme/skeleton/partials/_recordfooter.twig (hunk 2 above)
Comment (JarJak) [reply public/theme/skeleton/partials/_recordfooter.twig]: Renamed to related_content_types

File: src/Entity/Content.php (hunk 3 above)
Comment (JarJak) [reply src/Entity/Content.php]: fixed

File: src/Entity/Content.php (hunk 4 above)
Comment (JarJak) [reply src/Entity/Content.php]: fixed

**Pull request merged**